### PR TITLE
[SPARK-49086][CONNECT] Move ML function registration to SparkSessionExtensions

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR moves ML function registration from the SparkConnectPlanner to the internal function registry. This registration is done using the SparkSessionExtensions mechanism.

### Why are the changes needed?
Unification of Connect and Classic Column API. This PR decouples the ML functions from Catalyst expressions.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Existing tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]
cc @WeichenXu123 and @zhengruifeng

**Review by zhengruifeng — APPROVED** [review]

**Review by WeichenXu123 — APPROVED** [review]
LGTM.

File: mllib/src/main/scala/org/apache/spark/sql/ml/FunctionRegistration.scala [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import org.apache.spark.sql.expressions.UserDefinedFunction
+import org.apache.spark.sql.functions.udf
+
+/**
+ * Register a couple ML vector conversion UDFs in the internal function registry.
+ *
+ * This is a bit of a hack because we use the [[SparkSessionExtensions]] mechanism to register
+ * functions in a global registry. The use of a Scala object makes sure we only do this once.
+ */
+object FunctionRegistration {
```
Comment (allisonwang-db) [inline mllib/src/main/scala/org/apache/spark/sql/ml/FunctionRegistration.scala:34]: super nit: `InternalFunctionRegistration`?

**Review by allisonwang-db — APPROVED** [review]

**Comment by hvanhovell** [comment]
@zhengruifeng this is failing `org.apache.spark.ml.FunctionsLoadingSuite`. Can I remove this? I can't make it error out anymore after this.

**Comment by zhengruifeng** [comment]
@hvanhovell I think we can remove test `org.apache.spark.ml.FunctionsLoadingSuite` introduced in https://github.com/apache/spark/pull/43739, because we don't need objects `vectorToArrayUdf` and `vectorToArrayFloatUdf` any more

also cc @WeichenXu123 and @zsxwing who are the reviewers of https://github.com/apache/spark/pull/43739

File: mllib/src/main/scala/org/apache/spark/sql/ml/InternalFunctionRegistration.scala [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import org.apache.spark.sql.functions.udf
+import org.apache.spark.sql.internal.UserDefinedFunctionUtils.toScalaUDF
+
+/**
+ * Register a couple ML vector conversion UDFs in the internal function registry.
+ *
+ * This is a bit of a hack because we use the [[SparkSessionExtensions]] mechanism to register
+ * functions in a global registry. The use of a Scala object makes sure we only do this once.
+ */
+object InternalFunctionRegistration {
```
Comment (hvanhovell) [inline mllib/src/main/scala/org/apache/spark/sql/ml/InternalFunctionRegistration.scala:35]: @grundprinzip here

**Review by grundprinzip — APPROVED** [review]
I like it a lot. Thanks!

**Comment by HyukjinKwon** [comment]
@hvanhovell wanted to merge this but there's a conflict ..

**Comment by hvanhovell** [comment]
Merging to master!

**Pull request closed without merging**